src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/LocalDB.Windows.cs

```diff
@@ -0,0 +1,218 @@
+                }
+
+                _sqlUserInstanceLibraryHandle = libraryHandle;
+
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Retireves the part of the sqlUserInstance.dll from the registry
+        /// </summary>
+        /// <param name="errorState">In case the dll path is not found, the error is set here.</param>
+        /// <returns></returns>
+        private string GetUserInstanceDllPath(out LocalDBErrorState errorState)
+        {
+            string dllPath = null;
+            using (RegistryKey key = Registry.LocalMachine.OpenSubKey(LocalDBInstalledVersionRegistryKey))
+            {
+                if (key == null)
+                {
+                    errorState = LocalDBErrorState.NO_INSTALLATION;
+                    return null;
+                }
+
+                Version zeroVersion = new Version("0.0");
+
+                Version latestVersion = zeroVersion;
+
+                foreach (string subKey in key.GetSubKeyNames())
+                {
+                    try
+                    {
+                        Version currentKeyVersion = new Version(subKey);
+                        latestVersion = latestVersion.CompareTo(currentKeyVersion) < 0 ? currentKeyVersion : latestVersion;
+                    }
+                    catch (FormatException)
+                    {
+                        errorState = LocalDBErrorState.INVALID_CONFIG;
+                        return null;
+                    }
+                }
+
+                // If no valid versions are found, then error out
+                if (latestVersion.Equals(zeroVersion))
+                {
+                    errorState = LocalDBErrorState.INVALID_CONFIG;
+                    return null;
+                }
+
+                // Use the latest version to get the DLL path
+                using (RegistryKey latestVersionKey = key.OpenSubKey(latestVersion.ToString()))
+                {
+
+                    object instanceAPIPathRegistryObject = latestVersionKey.GetValue(InstanceAPIPathValueName);
+                    if (instanceAPIPathRegistryObject == null)
+                    {
+                        errorState = LocalDBErrorState.NO_SQLUSERINSTANCEDLL_PATH;
+                        return null;
+                    }
+
+                    RegistryValueKind valueKind = latestVersionKey.GetValueKind(InstanceAPIPathValueName);
+
+                    if (valueKind != RegistryValueKind.String)
+                    {
+                        errorState = LocalDBErrorState.INVALID_SQLUSERINSTANCEDLL_PATH;
+                        return null;
+                    }
+                    else
+                    {
+                        dllPath = (string)instanceAPIPathRegistryObject;
+                    }
+
+                    errorState = LocalDBErrorState.NONE;
+                    return dllPath;
+                }
+            }
+        }
+
+        ~LocalDB()
```

Comment:
@stephentoub what's the rule about when a finalizer is necessary (beyond just being disposable/having disposable types in your fields)